include/pagmo/algorithms/nsga2.hpp

```diff
@@ -0,0 +1,571 @@
+          m_verbosity(0u), m_log()
+    {
+        if (cr >= 1. || cr < 0.) {
+            pagmo_throw(std::invalid_argument, "The crossover probability must be in the [0,1[ range, while a value of "
+                                                   + std::to_string(cr) + " was detected");
+        }
+        if (m < 0. || m > 1.) {
+            pagmo_throw(std::invalid_argument, "The mutation probability must be in the [0,1] range, while a value of "
+                                                   + std::to_string(cr) + " was detected");
+        }
+        if (eta_c < 1. || eta_c >= 100.) {
+            pagmo_throw(std::invalid_argument,
+                        "The distribution index for crossover must be in [1, 100[, while a value of "
+                            + std::to_string(eta_c) + " was detected");
+        }
+        if (eta_m < 1. || eta_m >= 100.) {
+            pagmo_throw(std::invalid_argument,
+                        "The distribution index for mutation must be in [1, 100[, while a value of "
+                            + std::to_string(eta_m) + " was detected");
+        }
+    }
+
+    /// Algorithm evolve method (juice implementation of the algorithm)
+    /**
+     *
+     * Evolves the population for the requested number of generations.
+     *
+     * @param pop population to be evolved
+     * @return evolved population
+     * @throw std::invalid_argument if pop.get_problem() is stochastic, single objective or has non linear constraints.
+     * If \p int_dim is larger than the problem dimension. If the population size is smaller than 5 or not a multiple of
+     * 4.
+     */
+    population evolve(population pop) const
+    {
+        // We store some useful variables
+        const auto &prob = pop.get_problem(); // This is a const reference, so using set_seed for example will not be
+                                              // allowed (pop.set_problem_seed is)
+        auto dim = prob.get_nx();             // This getter does not return a const reference but a copy
+        auto NP = pop.size();
+
+        auto fevals0 = prob.get_fevals(); // discount for the fevals already made
+        unsigned int count = 1u;          // regulates the screen output
+
+        // PREAMBLE-------------------------------------------------------------------------------------------------
+        // We start by checking that the problem is suitable for this
+        // particular algorithm.
+        if (prob.is_stochastic()) {
+            pagmo_throw(std::invalid_argument,
+                        "The problem appears to be stochastic " + get_name() + " cannot deal with it");
+        }
+        if (prob.get_nc() != 0u) {
+            pagmo_throw(std::invalid_argument, "Non linear constraints detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them.");
+        }
+        if (prob.get_nf() < 2u) {
```

Comment:
This error message seems a bit needlessly complicated. Can't it be something like "this is a multiobjective algortihm and the problem prob.get_name() is single-objective"?